### PR TITLE
add SSR for legacy Ethereum proposals graph query

### PR DESCRIPTION
**Pull request by 0xMillz**
Deployed here: [gnars-frontend-f6f0slrw5-gnars.vercel.app](https://vercel.live/link/gnars-frontend-f6f0slrw5-gnars.vercel.app?via=deployment-domains-list-commit)

Visual difference for the code, Only 1 client-side request for proposals vs 6 in 15 seconds on the same `/dao` page:
image

I also realized "pending" still means "active" so fixed that along with a lot of other legacy code I ripped out

### CODE REVIEW

File: packages/webapp/constants/types.ts [hunk 1]
```diff
@@ -11,10 +11,10 @@ export const enum CHAIN_IDS {
   BASE_SEPOLIA = 84532
 }
 
-export const ALCHEMY_RPC_URLS: Map<number, string> = new Map([
-  [CHAIN_IDS.ETHEREUM, "https://eth-mainnet.g.alchemy.com/v2/"],
-  [CHAIN_IDS.BASE, "https://base-mainnet.g.alchemy.com/v2/"]
-])
+export const ALCHEMY_RPC_URLS= {
```
Comment (0xMillz) [inline packages/webapp/constants/types.ts:14]: Typescript likes this syntax better because `map.get(X)` can potential return `undefined`

File: packages/webapp/env/client.ts [hunk 2]
```diff
@@ -0,0 +1,6 @@
+export const alchemyApiKey = process.env.NEXT_PUBLIC_ALCHEMY_API_KEY
```
Comment (0xMillz) [inline packages/webapp/env/client.ts:1]: Split these into client & server env files so I could use @t3-oss/env-next

File: packages/subgraph/README.md [hunk 3]
```diff
@@ -67,10 +67,5 @@ A good practice to deploy in mainnet is to have a duplicate/backup subgraph so t
...
 
-## Current Deployments:
-
-### Mainnet
-
-API Endpoint: https://api.goldsky.com/api/public/project_clwh32mvalyh201vi1ck71seq/subgraphs/gnars-mainnet/latest/gn
-
-Explorer: https://api.goldsky.com/api/public/project_clwh32mvalyh201vi1ck71seq/subgraphs/gnars-mainnet/latest/gn
+## Current Deployments:  
+Goldsky, Alchemy, and The Graph decentralized network
```
Comment (0xMillz) [inline packages/subgraph/README.md:71]: I ran out of GRT to pay for my Graph API Key so both our graphs are pointed at Goldsky now in the code and config I pushed. I would have moved to Alchemy for at least one of them but after 16 hours it was only 75% indexed

File: packages/webapp/.env.example [hunk 4]
```diff
@@ -5,4 +5,15 @@ NEXT_PUBLIC_BASESCAN_API_KEY=XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX
 NEXT_PUBLIC_WALLET_CONNECT_PROJECT_ID=XXXXXXXXXXXXXXXXXXXXXXXXXXXXX
```
Comment (0xMillz) [inline packages/webapp/.env.example:5]: We'll have some redundant env options to choose which subgraphs to use untill we settle on a final solution. I put all the values in prod vercel so any of these variables work as long as your code is pointing to the one you want

File: packages/webapp/.eslintrc.json [hunk 5]
```diff
@@ -1,6 +1,6 @@
 {
   "extends": [
-    "next/core-web-vitals",
-    "plugin:storybook/recommended"
```
Comment (0xMillz) [inline packages/webapp/.eslintrc.json:4]: This legacy version of Storybook was really getting in the way progress, especially with TS, so I just ripped it al out for now. Version 7 and 8 are completely new re-writes so maybe we can check them out in the future.

File: packages/webapp/components/Governance/Delegation/UpdateDelegateModal.tsx [hunk 6]
```diff
@@ -52,7 +52,7 @@ export const UpdateDelegateModal: FC<UpdateDelegateModalProps> = ({ onClose, ...
     } catch {
       setIsValidName.off()
     }
-  }, [accountQuery])
+  }, [accountQuery, setIsValidName])
```
Comment (0xMillz) [inline packages/webapp/components/Governance/Delegation/UpdateDelegateModal.tsx:55]: Every seeing this warning > 100 times running `pnpm dev` it had to be fixed

**Comment by asmattic** [comment]
image

When I click DAO on the top nav there's a serious error loop. This is after only a few seconds.

**Comment by bagelface** [comment]
> image
> @0xMillz
> 
> When I click DAO on the top nav there's a serious error loop. This is after only a few seconds, it was climbing fast.

any steps to recreate this issue? not seeing it on the vercel deployment. is this only when you run it locally?

File: packages/webapp/next.config.mjs [hunk 7]
```diff
@@ -1,7 +1,28 @@
...
+
+switch (VERCEL_ENV) {
+  case "preview":
+    ROOT_URI = `https://${VERCEL_URL}/api/graphql`;
+    break;
+  case "production":
+    ROOT_URI = `https://gnars.wtf/api/graphql`;
+    break;
+  default:
+    ROOT_URI = `http://localhost:3000/api/graphql`;
```
Comment (asmattic) [inline packages/webapp/next.config.mjs:16]: detect the port dynamically used from the `env` vars. If something is already running on port `3000` it will default to `3001`, etc.

File: packages/webapp/pages/dao/index.tsx [hunk 8]
```diff
@@ -21,47 +20,27 @@ import { getProposals, ProposalsResponse } from "@queries/base/requests/proposal
...
-
   const { data: baseData, error: baseError } = useQuery(
     [USE_QUERY_KEYS.PROPOSALS, CHAIN_IDS.BASE, tokenAddress, page],
     () => getProposals(CHAIN_IDS.BASE, tokenAddress, LIMIT),
     {
       enabled: baseQueryReady
     }
   )
+
   if (baseError) {
```
Comment (asmattic) [inline packages/webapp/pages/dao/index.tsx:40]: What happens in the UI if there is a base error?

File: packages/webapp/constants/types.ts (hunk 1 above)
Comment (bagelface) [inline packages/webapp/constants/types.ts:14]: tbh this and CHAIN_IDS doesn't seem like it belongs in a "types" file

File: pnpm-workspace.yaml
Comment (asmattic) [inline pnpm-workspace.yaml:1]: Why was this deleted?

File: pnpm-lock.yaml
Comment (asmattic) [inline pnpm-lock.yaml:1]: Why was this deleted? Didn't we decide we want to track lockfile changes

File: packages/webapp/pages/dao/index.tsx
Comment (asmattic) [inline packages/webapp/pages/dao/index.tsx:1]: I'm going to test locally for these changes. I'll resolve or have comments after.

File: packages/webapp/pages/api/graphql.ts
Comment (asmattic) [inline packages/webapp/pages/api/graphql.ts:1]: I'm going to test locally for these changes. I'll resolve or have comments after.

**Review by asmattic — CHANGES_REQUESTED** [review]
I have some requested changes and clarifications. ~~Going to test locally and confirm that the error loop I was getting is only my machine/browser and not intermittent.~~ 

Found the error was with the Vercel highlight and create thread widget.

File: packages/webapp/.env.example [hunk 9]
```diff
@@ -5,4 +5,15 @@ NEXT_PUBLIC_BASESCAN_API_KEY=XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX
...
+
+NEXT_PUBLIC_GOLDSKY_BASE_API_URL=https://api.goldsky.com/api/public/project_clz4ukquribdy010b1fgua9nm/subgraphs/gnars-base/latest/gn
+
+# Only using this on SSR for right now, we can make a NEXT_PUBLIC_ envar later one if needed
+PRIVATE_GOLDSKY_L1_SUBGRAPH=https://api.goldsky.com/api/public/project_clwh32mvalyh201vi1ck71seq/subgraphs/gnars-mainnet/latest/gn
+
+# only needed if we are using The Graph, but API usage costs $$ or GRT or you get rate blocked
+PRIVATE_GRAPH_LAYER_1_API_KEY='YOUR PERSONAL API KEY (FOR THE GRAPH NETWORK)'
+
+#TODO: implement and actually make private (ignore for now)
```
Comment (bagelface) [inline packages/webapp/.env.example:18]: will make a separate issue for this

File: packages/webapp/graphql/ssr.client.ts [hunk 10]
```diff
@@ -0,0 +1,4 @@
+import { GraphQLClient } from 'graphql-request';
+
+// this hits the graphql api exposed in the pages/api/graphql.ts file
+export const graphQLClient = new GraphQLClient(process.env.ROOT_URI as string);
```
Comment (bagelface) [inline packages/webapp/graphql/ssr.client.ts:4]: ROOT_URI should have a better name and be included in the .env.example file

File: packages/webapp/next.config.mjs [hunk 11]
```diff
@@ -1,7 +1,28 @@
+const {
+  VERCEL_ENV = 'development',
+  VERCEL_URL = 'localhost'
+} = process.env
```
Comment (bagelface) [inline packages/webapp/next.config.mjs:4]: should probably be included in .env.example

**Comment by asmattic** [comment]
> > image
> > @0xMillz
> > When I click DAO on the top nav there's a serious error loop. This is after only a few seconds, it was climbing fast.
> 
> any steps to recreate this issue? not seeing it on the vercel deployment. is this only when you run it locally?

I found the issue. It's the frontend `start a thread` tool. If you highlight something and don't start a thread it will rack up those errors until exited out. So nothing to do with the PR changes.

File: packages/webapp/utils/web3.ts [hunk 12]
```diff
@@ -1,4 +1,15 @@
...
+import { alchemyApiKey } from "@env/client.ts"
+import { Block, createPublicClient, http } from "viem"
+import { mainnet } from 'viem/chains';
+
+const viemClient = createPublicClient({
+  chain: mainnet,
+  transport: http(ALCHEMY_RPC_URLS[CHAIN_IDS.ETHEREUM] + alchemyApiKey as string),
+});
+
+import { ALCHEMY_RPC_URLS, CHAIN_IDS } from "@constants/types.ts"
```
Comment (bagelface) [inline packages/webapp/utils/web3.ts:11]: move to be with other imports

File: packages/webapp/pages/dao/index.tsx [hunk 13]
```diff
@@ -174,3 +153,52 @@ export default function Proposals() {
...
+      props: {
+        ethProposals: [[], []]
+      }
+    }
+  }
+}
+
+// Ethereum proposals only
+async function fetchProposalsData(queryClient: QueryClient, block?: Block): Promise<any> {
+  return await queryClient.fetchQuery([USE_QUERY_KEYS.PROPOSALS, block?.number?.toString()], fetchProposals)
```
Comment (bagelface) [inline packages/webapp/pages/dao/index.tsx:178]: redundant return await

File: packages/webapp/pages/dao/index.tsx [hunk 14]
```diff
@@ -174,3 +153,52 @@ export default function Proposals() {
...
+  )
+}
+
+async function fetchEthProposals(block?: Block): Promise<any[]> {
+  const queryClient = new QueryClient()
+  const proposalsData = await fetchProposalsData(queryClient, block)
+  const proposals = mapProposals(proposalsData, block)
+  const [activeProposals, finalizedProposals] = partition(proposals, (p) => !isFinalized(p.effectiveStatus))
+
+  return [ activeProposals, finalizedProposals]
```
Comment (bagelface) [inline packages/webapp/pages/dao/index.tsx:203]: remove extra space

File: packages/webapp/pages/dao/index.tsx [hunk 15]
```diff
@@ -174,3 +153,52 @@ export default function Proposals() {
...
+  const block: Block | undefined = await getLatestBlock()
+  try {
+    const ethProposals  = await fetchEthProposals(block)
+    return {
+      props: {
+        ethProposals,
+      }
+    }
+  } catch (err) {
+    console.error('Error getting Eth Proposals from subgraph', err)
```
Comment (bagelface) [inline packages/webapp/pages/dao/index.tsx:167]: Error getting Ethereum proposals from subgraph

File: packages/webapp/next.config.mjs (hunk 11 above)
Comment (asmattic) [reply packages/webapp/next.config.mjs:4]: Agreed. Though I'd say we're at the point where we need to break out specific localhost dev vs anything on Vercel vars in  `.env.example` and `env.local.example`

**Review by bagelface — CHANGES_REQUESTED** [review]
see comments

File: packages/webapp/constants/types.ts (hunk 1 above)
Comment (asmattic) [reply packages/webapp/constants/types.ts:14]: I thought, and think the same. It's how it's set up in `nouns-builder` so it makes all the dozens and dozens of import changes easier while getting all the other connections working. I've been using it the same way with auctions. I think once auctions/proposals are in working order we can revisit with a find replace, but is worth it to keep as is for now. Can you create an issue for revisiting?

File: packages/webapp/.env.example (hunk 9 above)
Comment (asmattic) [reply packages/webapp/.env.example:18]: Relevant to this topic: (reference to my comments on `.env.example` and `env.local.example` below)[https://github.com/gnars-dao/gnars-frontend/pull/48#discussion_r1718785256]

**Comment by 0xMillz** [comment]
> image
> @0xMillz
> 
> When I click DAO on the top nav there's a serious error loop. This is after only a few seconds, it was climbing fast.

I noticed it once, and I think I traced coming from my MetaMask extension

**Comment by bagelface** [comment]
> > image
> > @0xMillz
> > When I click DAO on the top nav there's a serious error loop. This is after only a few seconds, it was climbing fast.
> 
> I noticed it once, and I think I traced coming from my MetaMask extension

probably just an RPC error then

File: packages/webapp/constants/stuff.ts
Comment (bagelface) [inline packages/webapp/constants/stuff.ts:1]: would name this something like "environment.ts", "network.ts", "chains.ts"

File: packages/webapp/data/contract/requests/getDAOAddresses.ts [hunk 16]
```diff
@@ -1,7 +1,8 @@
 import { readContract } from 'wagmi/actions'
 
 import { NULL_ADDRESS, PUBLIC_MANAGER_ADDRESS } from '@constants/baseAddresses.ts'
-import { AddressType, CHAIN_IDS } from '@constants/types'
+import { AddressType} from '@constants/types'
```
Comment (bagelface) [inline packages/webapp/data/contract/requests/getDAOAddresses.ts:4]: need space after "AddressType"

File: packages/webapp/graphql/ssr.client.ts [hunk 17]
```diff
@@ -0,0 +1,4 @@
+import { GraphQLClient } from 'graphql-request';
+
+// this hits the graphql api exposed in the pages/api/graphql.ts file
```
Comment (bagelface) [inline packages/webapp/graphql/ssr.client.ts:3]: Don't think this comment is necessary tbh, it's sort of self commenting. I also do not like comments that reference specific implementation details, in this case the path of the API, because it's very easy for them to go out of date unnoticed and cause more confusion than they relieve in the future.

File: packages/webapp/pages/dao/index.tsx [hunk 18]
```diff
@@ -174,3 +153,52 @@ export default function Proposals() {
     </DarkMode>
   )
 }
+
+export async function getServerSideProps() {
+  const block: Block | undefined = await getLatestBlock()
```
Comment (bagelface) [inline packages/webapp/pages/dao/index.tsx:158]: might be good if this function indicated it's getting the latest ethereum block. ambiguous as it currently is

**Review by bagelface — CHANGES_REQUESTED** [review]
a few small suggestions

File: packages/webapp/pages/dao/index.tsx (hunk 18 above)
Comment (0xMillz) [reply packages/webapp/pages/dao/index.tsx]: 👍

File: packages/webapp/constants/stuff.ts
Comment (0xMillz) [reply packages/webapp/constants/stuff.ts:1]: I went with the TypeScript convention and am exporting everything from in constants from its `index.ts` file

**Review by bagelface — CHANGES_REQUESTED** [review]
mixed use of single and double quotes throughout

**Review by bagelface — APPROVED** [review]
lgtm

File: packages/webapp/next.config.mjs (hunk 11 above)
Comment (asmattic) [reply packages/webapp/next.config.mjs:5]: I know this is already merged, but as a correction to this @0xMillz. You are right. I believe I put in a different comment response what you said here, that those env vars are already on `process.env.{AUTO_GENERATED_ENV_VAR}`.
